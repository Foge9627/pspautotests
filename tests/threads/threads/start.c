#include "shared.h"

volatile int argLengthArgc = 0;
volatile void *argLengthArgv;
int argLengthFunc(SceSize argc, void *argv) {
	argLengthArgc = argc;
	argLengthArgv = argv;
	return 0;
}

int testFunc(SceSize argc, void *argv) {
	schedf("* delayFunc\n");

	sceKernelDelayThread(500);

	return 7;
}

int main(int argc, char *argv[]) {
	int i, result;
	SceUID testThread = sceKernelCreateThread("test", &testFunc, sceKernelGetThreadCurrentPriority(), 0x1000, 0, NULL);
	SceUID deletedThread = sceKernelCreateThread("deleted", &testFunc, sceKernelGetThreadCurrentPriority(), 0x1000, 0, NULL);
	sceKernelDeleteThread(deletedThread);

	checkpointNext("Thread IDs:");
	checkpoint("  Normal: %08x", sceKernelStartThread(testThread, 0, NULL));
	checkpoint("  Twice: %08x", sceKernelStartThread(testThread, 0, NULL));
	checkpoint("  NULL: %08x", sceKernelStartThread(0, 0, NULL));
	checkpoint("  Current: %08x", sceKernelStartThread(sceKernelGetThreadId(), 0, NULL));
	checkpoint("  Deleted: %08x", sceKernelStartThread(deletedThread, 0, NULL));
	checkpoint("  Invalid: %08x", sceKernelStartThread(0xDEADBEEF, 0, NULL));

	checkpointNext("Argument Length:");
	SceUID argLengthThread = sceKernelCreateThread("argLength", &argLengthFunc, sceKernelGetThreadCurrentPriority() - 1, 0x800, 0, NULL);
	char argLengthTemp[0x1000];

	result = sceKernelStartThread(argLengthThread, 8, NULL);
	sceKernelWaitThreadEnd(argLengthThread, NULL);
	checkpoint("  With NULL ptr: %08x (%d, %s)", result, argLengthArgc, argLengthArgv ==  0 ? "NULL" : (argLengthTemp == argLengthTemp ? "same ptr" : "unknown"));

	// Note: larger than stack crashes the PSP, and so does a bad arg ptr.
	static int lengths[] = {-1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 80, 90, 0x600};
	for (i = 0; i < ARRAY_SIZE(lengths); ++i) {
		argLengthArgc = -1;
		result = sceKernelStartThread(argLengthThread, lengths[i], argLengthTemp);
		sceKernelWaitThreadEnd(argLengthThread, NULL);
		checkpoint("  %d arg length: %08x (%d, %s)", lengths[i], result, argLengthArgc, argLengthArgv ==  0 ? "NULL" : (argLengthTemp == argLengthTemp ? "same ptr" : "unknown"));
	}

	checkpointNext("Priorities:");
	static int priorities[] = {0x8, 0x1F, 0x20, 0x21, 0x40};
	for (i = 0; i < ARRAY_SIZE(priorities); ++i) {
		argLengthArgc = -1;
		sceKernelDeleteThread(argLengthThread);
		argLengthThread = sceKernelCreateThread("argLength", &argLengthFunc, priorities[i], 0x800, 0, NULL);
		result = sceKernelStartThread(argLengthThread, priorities[i], argLengthTemp);
		int afterStart = argLengthArgc;
		sceKernelWaitThreadEnd(argLengthThread, NULL);

		int priorityDiff = priorities[i] - sceKernelGetThreadCurrentPriority();
		checkpoint("  priority 0x%02x: %08x (current%s0x%02x, %s)", priorities[i], result, priorityDiff < 0 ? "-" : "+", priorityDiff < 0 ? -priorityDiff : priorityDiff, afterStart == priorities[i] ? "resched" : (argLengthArgc == priorities[i] ? "deferred" : "never"));
	}

	return 0;
}