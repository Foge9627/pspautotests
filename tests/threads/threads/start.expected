[x] Thread IDs:
[x]   Normal: 00000000
[x]   Twice: 800201a4
[x]   NULL: 80020197
[x]   Current: 800201a4
[x]   Deleted: 80020198
[x]   Invalid: 80020198

[x] Argument Length:
[x]   With NULL ptr: 00000000 (0, NULL)
[x]   -1 arg length: 800200d3 (-1, NULL)
[x]   0 arg length: 00000000 (0, NULL)
[x]   1 arg length: 00000000 (1, same ptr)
[x]   2 arg length: 00000000 (2, same ptr)
[x]   3 arg length: 00000000 (3, same ptr)
[x]   4 arg length: 00000000 (4, same ptr)
[x]   5 arg length: 00000000 (5, same ptr)
[x]   6 arg length: 00000000 (6, same ptr)
[x]   7 arg length: 00000000 (7, same ptr)
[x]   8 arg length: 00000000 (8, same ptr)
[x]   80 arg length: 00000000 (80, same ptr)
[x]   90 arg length: 00000000 (90, same ptr)
[x]   1536 arg length: 00000000 (1536, same ptr)

[x] Priorities:
[x]   priority 0x08: 00000000 (current-0x18, resched)
[x]   priority 0x1f: 00000000 (current-0x01, resched)
[r]   priority 0x20: 00000000 (current+0x00, deferred)
[r]   priority 0x21: 00000000 (current+0x01, deferred)
[r]   priority 0x40: 00000000 (current+0x20, deferred)
